### ISSUE TITLE
search doesn't work as expected (e.g. search for "doormat" doesn't return "keys under doormats" paper)

### ISSUE THREAD

**Issue by dkg** [The-Encryption-Compendium/TECv2#25]
Consider the query for "doormat": https://encryptioncompendium.org/search/?query=doormat

I would have expected this to return a list with at least [the "Keys Under Doormats" paper from 2015](https://encryptioncompendium.org/entries/mandating-insecurity-by-requiring-government-acces-5a5c309b/).  Indeed, [data.bib](https://raw.githubusercontent.com/The-Encryption-Compendium/the-encryption-compendium.github.io/master/data/data.bib), which i think is the source for these items, lists that paper first.

However, the only thing i see as a result is 1994's "Information Security and Privacy in Network Environments", which i don't think refers to doormats at all:

![doormat-search](https://user-images.githubusercontent.com/66738/133506617-98937c0a-b4e2-44af-a32a-5ee441c8992d.png)

The same thing happens at https://the-encryption-compendium.github.io/search/?query=doormat as far as i can tell (i don't understand the difference between the two sites, though the path in the URLs ultimately differ between the two).

**Comment by dkg** [comment]
another search that doesn't work as expected is a [search for "Diffie"](https://encryptioncompendium.org/search/?query=Diffie) -- this yields 2 results ("The Risks of Key Recovery, Key Escrow, and Trusted Third-Party Encryption" and "New Directions in Cryptography"), but "doormats" (which Diffie collaborated on) is not listed.

Even weirder, a [search for "Diﬃe" (note the use of `U+FB03 LATIN SMALL LIGATURE FFI` instead of the three-character string "ffi")](https://encryptioncompendium.org/search/?query=Diffie) lists "New Directions in Cryptography" but doesn't list "Doormats" or "The Risks of Key Recovery, Key Escrow, and Trusted Third-Party Encryption (Revised)" (which Diffie also co-authored) -- yet three new results show up (Wray's "The Way Forward: Working Together to Tackle Cybercrime", Comey's "Going Dark: Are Technology, Privacy, and Public Safety on a Collision Course?", and EFF's "Deep Dive into Crypto “Exceptional Access” Mandates: Effective or Constitutional—Pick One") none of which even mention Diffie afaict.  (i tested with the ligature because `data.bib` uses the ligature to represent Diffie's name in the bibtex entry for "Doormats")

**Comment by kernelmethod** [comment]
Thanks for looking into this and other issues on the site, it's a big help! These is definitely some weird behavior to be seeing. First, re: your question

> i don't understand the difference between the two sites, though the path in the URLs ultimately differ between the two

In theory there shouldn't be any difference between the sites, but I've made some small improvements to [encryptioncompendium.org](https://encryptioncompendium.org) that have yet to be merged into master. I just opened #30 to get those changes into master, and after that they should be identical.

As for your question here, after some investigation I think that the issues you're seeing are mostly coming from 2-3 sources:

1. Various quirks in the way that compendium entries were originally entered into Zotero as well as how Zotero exports them (although I admit that I truly do not understand how `"Diffie"` got stored as `"Dif\ufb03e"` :frowning_face:);
2. Hacks in the [`generate_compendium.py`](https://github.com/The-Encryption-Compendium/TECv2/blob/master/deploy_tools/generate_compendium.py) script that'll occasionally cause it to do things like miss authors in a compendium entry; and
3. Bad [Fuse.js configuration options](https://fusejs.io/api/options.html) set in [`search.js`](https://github.com/The-Encryption-Compendium/TECv2/blob/master/assets/js/search.js).

So for the particular queries you mentioned,

- "doormat" isn't showing the "Keys Under Doormats" paper because it appears that only the second half of the title of that paper was recorded as its title, while "Keys Under Doormats" was entered into a separate "shorttitle" field. I'm not completely certain why "Information Security and Privacy in Network Environments" is showing up, but after some experiments I suspect it's an issue with the version of Fuse.js that we're using as well as with the parameters I've selected (since Fuse.js does a fuzzy search on the compendium entries).
- The "Keys Under Doormats" paper isn't showing up for the query "diffie" because Fuse.js will only search a part of a string for a substring, which is dictated by the "location", "distance", and "threshold" parameters. The authors list for that paper is long enough that it didn't actually see Diffie as one of the authors of the paper.
- I can't quite explain yet why the `"Dif\ufb03e"` query is giving the results that it is, but my guess is that it's related again to the fact that Fuse.js is doing a fuzzy search.

**Comment by kernelmethod** [comment]
I've fixed a bug in the `generate_compendium.py` script in #30 that should partially fix problem (2) that I listed above, and as for problem (3), I'll need to create a separate PR to update the Fuse.js version as well as the parameters. So far, it looks like turning down `threshold` a little (e.g. to `0.1`) and either making `distance` very large (e.g. `"distance": 1000`) or setting `"ignoreLocation": true` in `search.js` should resolve a lot of issues.

As for problems in the entries itself, there should be a large update coming to the compendium fairly soon to help clean up issues like that (as well as some other substantial improvements). Unfortunately there are still some weird things like `"Diffie"` being stored as `"Dif\ufb03e"` that still persist, plus some issues with Zotero and the way it exports data. I can make some of those changes manually (I've already gone in and fixed the title on the "Keys Under Doormat" paper as well as the Diffie Unicode problem, although those changes still need to be propagated to the site), but admittedly there's still a lot more work to be done on that front.

**Comment by dkg** [comment]
hm, now [searching for "rethinking"](https://the-encryption-compendium.github.io/search/?query=rethinking) doesn't return anything on the github.io staging site.  I think search is fully broken there now.  (i'd expect it to show at least "rethinking encryption" by Jim Baker)

**Comment by kernelmethod** [comment]
Woops -- I borked the GitHub Pages site by adding some `.gitignore` files that got copied over into that repository. It should be fixed now!

**Comment by dkg** [comment]
thanks for that fix.  I note that with the changed parameters, a query for "random" [on the staging site](https://the-encryption-compendium.github.io/search/?query=random) now includes "Memorandum for the Heads of Executive Departments and Agencies: Policy to Require Secure Connections Across Federal Websites and Web Services" whereas [the same query on the live site](https://encryptioncompendium.org/search/?query=random) does not.  i think it's going from "random" to "memorandum" which is maybe a little too loose, no?

**Comment by kernelmethod** [comment]
> i think it's going from "random" to "memorandum" which is maybe a little too loose, no?

Yeah, it is. Let me try tweaking the Fuse.js parameters a bit and see if I can get that fixed today!

**Comment by kernelmethod** [comment]
I've just updated the Fuse.js parameters and brought it to the latest version in #50, and from my tests it seems like all of the queries you've mentioned here have been fixed! I'm going to keep this open for now just in case any more weird search behavior crops up, but I think the situation should be a lot better than it was before.
